File: backend/app/models/product.py
from datetime import datetime
from decimal import Decimal

from sqlalchemy import (
    Boolean,
    DateTime,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column

from app.database.base import Base


class Product(Base):
    __tablename__ = "products"

    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        index=True,
    )

    product_code: Mapped[str] = mapped_column(
        String(30),
        unique=True,
        nullable=False,
        index=True,
    )

    product_name: Mapped[str] = mapped_column(
        String(200),
        nullable=False,
    )

    description: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    category: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    unit: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
    )

    hsn_code: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
    )

    gst_percentage: Mapped[Decimal] = mapped_column(
        Numeric(5, 2),
        nullable=False,
    )

    purchase_price: Mapped[Decimal] = mapped_column(
        Numeric(12, 2),
        nullable=False,
    )

    selling_price: Mapped[Decimal] = mapped_column(
        Numeric(12, 2),
        nullable=False,
    )

    minimum_stock: Mapped[int] = mapped_column(
        Integer,
        default=0,
    )

    maximum_stock: Mapped[int] = mapped_column(
        Integer,
        default=0,
    )

    current_stock: Mapped[int] = mapped_column(
        Integer,
        default=0,
    )

    is_active: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False,
    )

    created_by: Mapped[int] = mapped_column(
        ForeignKey("users.id"),
        nullable=False,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        server_default=func.now(),
    )

    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        server_default=func.now(),
        onupdate=func.now(),
    )

File: frontend/src/types/product.ts
export interface Product {
    id: number;
  
    product_code: string;
    product_name: string;
  
    description: string | null;
  
    category: string;
    unit: string;
    hsn_code: string;
  
    gst_percentage: string;
    purchase_price: string;
    selling_price: string;
  
    minimum_stock: string;
    maximum_stock: string;
    current_stock: string;
  
    is_active: boolean;
  
    created_at: string;
    updated_at: string;
  }
  
  
  export interface ProductCreatePayload {
    product_code: string;
    product_name: string;
  
    description: string | null;
  
    category: string;
    unit: string;
    hsn_code: string;
  
    gst_percentage: number;
    purchase_price: number;
    selling_price: number;
  
    minimum_stock: number;
    maximum_stock: number;
    current_stock: number;
  
    is_active: boolean;
  }
  
  
  export type ProductUpdatePayload =
    ProductCreatePayload;